### PR TITLE
Update: Grouping related CLI options. (fixes #3612)

### LINKED ISSUE #3612: CLI options could be grouped and otherwise presented better
The output of `eslint --help` is pretty hard to read with the number of options we now have, whether or not the options have short or long descriptions.  Plus, the options are not in a clear order that anyone on the project could easily perceive, understand, and explain.

It might be a good idea to have a discussion on how we can improve the options listing.

I believe the single biggest improvement we could make is to group related options under separate headers. This will result in natural whitespace between groups, which would allow us to slightly relax UX standards for the individual option descriptions themselves. In addition, we could revisit some of the descriptions if necessary.

### PR DESCRIPTION
**Pull request by platinumazure**
With apologies to `git` for forcing a rewrite due to a crazy diff, here is my take on grouping CLI options.

Force-pushed an update right away because I wanted to add some line breaks around the sections. Sorry for any confusion.

Please weigh in with ordering or re-grouping suggestions.

Fixes #3612.

### CODE REVIEW

**Comment by platinumazure** [comment]
One open question still on the issue:

> One more question for the group: Should we keep docs/user-guide/command-line-interface.md exactly in sync with the options?
> 
> Or, more specifically:
> 1. Should I group the options there, in the same way as I am doing for lib/options.js?
> 2. Do we really want exactly the optionator text there, when perhaps it might be sufficient to have just the options and the basic description there and let the rest of the page flesh out details and defaults?

File: lib/options.js [hunk 1]
```diff
@@ -19,121 +19,159 @@ module.exports = optionator({
...
-        description: "Run config initialization wizard"
-    },
-    {
-        option: "parser",
-        type: "String",
-        default: "espree",
-        description: "Specify the parser to be used"
-    }]
+    options: [
+
```
Comment (ilyavolodin) [inline lib/options.js]: Extra line

File: lib/options.js [hunk 2]
```diff
@@ -19,121 +19,159 @@ module.exports = optionator({
...
+            option: "plugin",
+            type: "[String]",
+            description: "Specify plugins"
+        },
+        {
+            option: "rule",
+            type: "Object",
+            description: "Specify rules"
+        },
+
```
Comment (ilyavolodin) [inline lib/options.js]: Same

**Comment by ilyavolodin** [comment]
I think documentation should mirror help options. Would make it easier to read too. Also, there are a few more extra empty lines in the code in random places that I didn't mark.

**Comment by IanVS** [comment]
It looks like he's using the extra lines to visually separate the headings.

**Comment by platinumazure** [comment]
@ilyavolodin Yes, @IanVS has it right. Extra line breaks are for separating the headings and improving readability. According to the [coding conventions][1], there aren't any guidelines on arrays, but in objects, extra line breaks may be added to separate sections of an object for readability.

I'm tempted to make it a double line break so it's more clearly intentional. Would that be better?

  [1]: http://eslint.org/docs/developer-guide/code-conventions.html

**Comment by nzakas** [comment]
I don't think the extra lines add anything, so I'd vote for not having them.

Also agree, the docs should reflect this output.

**Comment by platinumazure** [comment]
@nzakas @ilyavolodin Updates have been made (removing blank lines and updating docs). I reordered the long descriptions to match the new grouping, but without adding subgroups-- let me know if you would like anything changed there.

**Comment by nzakas** [comment]
Looks nice. :) I think group order is fine, but we should have subheadings that correspond to the groups in the console output, otherwise people will complain that it looks random (even though it's not).

**Comment by platinumazure** [comment]
In that case, should the options become h4 and I should use h3 for the
group headers? (H2 is used in Ignoring Files subsequent header)

**Comment by nzakas** [comment]
Yup, sounds good to me

**Comment by platinumazure** [comment]
@nzakas Groupings have been added. Please let me know if anything else needs changing.

**Comment by nzakas** [comment]
This looks great! Thanks so much for the suggestion and the work to implement it.

**Pull request merged**